server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java

```diff
@@ -1463,7 +1463,7 @@ private void validateDataStreams(SortedMap<String, IndexAbstraction> indicesLook
                         List<String> conflicts = new ArrayList<>();
                         for (Map.Entry<String, IndexAbstraction> entry : potentialConflicts.entrySet()) {
                             if (entry.getValue().getType() != IndexAbstraction.Type.CONCRETE_INDEX ||
-                                indexNames.contains(entry.getKey()) == false) {
+                                indexNames.stream().noneMatch(name -> name.endsWith(entry.getKey()))) {
```

Comment:
we still need to be lenient here as, in the case where we're replacing a backing index with one that has a prefix, the source backing index will be detected as a conflict (it'll usually be the case, thinking of shrinking and searchable snapshot indices, that we'll promptly delete the source backing index so this is a temporary state)

Reply:
Wouldn't an index with a prefix not be identified as a potential conflict since it no longer starts with `ds.getName()`?

Reply:
This code is a bit difficult to read (in that it's difficult for me to understand what we're validating, and also exactly what we're checking for), I think some comments would be really helpful to signal behavior and intent, can you add comments for this (especially so it doesn't get reverted in the future)?

Reply:
I agree completely. 

So the reason why this needed relaxing was that we're setting a new `DataStream` after we replaced the original index. So from the `Metadata`'s perspective we're trying to add a `DataStream` with backing indices 
```
[shrink-foo-000001, foo-000002]
```
while the index we replaced `foo-000001` is not part of any `DataStream` anymore. 

@danhermann the `potentialConflicts` will contain `foo-000001` because the `ds.getName` in this case is `foo`.

I believe we have 2 options moving forward. 
1. keep this constraint relaxed and document the why much better
2. replace the way we replace a backing index by updating the `ClusterState` in 2 steps (note that I'm *not* saying we should do two cluster state updates, but rather locally build the `ClusterState` object twice [here](https://github.com/elastic/elasticsearch/pull/57616/files#diff-3052c7e3b380ea358d81ef172b9f43a7R90-R92). First step would be to add the `shrink-foo-000001` to the backing indices so after the first step the `DataStream` would look like 
```
[foo-000001, shrink-foo-000001, foo-000002]
```
and then remove index `foo-000001` from the backing indices.

What do you think? @danhermann @dakrone

Reply:
@andreidan, ah, thanks for the explanation. The original purpose of that validation was to prevent name collisions on rollover as you suggested. I have a PR out (https://github.com/elastic/elasticsearch/pull/57721) that changes the naming convention and I believe that will resolve this issue.

Reply:
Hmm, even with the naming change above, we'll still have the problem since ILM may be removing an index with the new naming scheme. I'll have to think about this one.

Reply:
@danhermann Thanks for pointing out [#57721](https://github.com/elastic/elasticsearch/pull/57721). I believe this problem will persist even with this PR in place. 

So after 57221 is merged, shrinking a backing index of a DataStream will atempt to create a DataStream with backing indices
```
[shrink-.ds-foo-000001, .ds-foo-000002]
```
while, at least temporary, index `ds-foo-000001` will still be found in the index lookup (as we're deleting it in a later step). 
Unless I'm missing something I think we still need to keep the updated validation as shown in this PR (or a different/better way? I'm a bit out of ideas on this one)

I just pushed some documentation for the validation to (hopefully) make reasoning about it easier

Reply:
Maybe we can change the validation to look at the names in `potentialConflicts` and flag them only if they would conflict  with a future rollover, i.e., with a number > the current generation.

Reply:
That's an interesting idea. Shall we do this in a separate PR to keep the scope of this (already rather large) PR manageable?

Reply:
Opened https://github.com/elastic/elasticsearch/pull/57750 for the validation change

Reply:
#57750 has been merged to master and I just merged master into this branch